### PR TITLE
Exchange the token exposed through environment variables before using it

### PR DESCRIPTION
**Pull request by pepicrft**
### WHY are these changes introduced?
We were trying to use the token generated by partners directly and it didn't work. The reason being is that the token has the right audience, `partners`, but the client, `partners`, can't be use from a different client `shopify-cli`.

### WHAT is this pull request doing?
This PR changes the workflow to exchange the token before using it. That'll give us a token that can be used from the `shopify-cli`.

### How to test your changes?
1. Authenticate with the CLI.
2. Open a console with `bin/console`.
3. Run `ShopifyCLI::IdentityAuth.new(ctx: ShopifyCLI::Context.new).exchange_partners_auth_token(ShopifyCLI::DB.new.get(:identity_access_token))`
4. You should get a refreshed token the current local `:identity_access_token`.

### Update checklist

- [x] I've added a CHANGELOG entry for this PR (if the change is public-facing)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows).
- [x] I've left the version number as is (we'll handle incrementing this when releasing).
- [ ] I've included any post-release steps in the section above.

### CODE REVIEW

File: lib/shopify_cli/identity_auth.rb [hunk 1]
```diff
@@ -195,30 +205,37 @@ def refresh_exchange_tokens
...
+        additional_scopes: additional_scopes,
+        subject_token: store.get(:identity_access_token),
+        destination: (name == "shopify") ? "https://#{store.get(:shop)}/admin" : nil
+      )
+      store.set("#{name}_exchange_token".to_sym => access_token)
+      ctx.debug("#{name}_exchange_token: " + access_token)
+    end
 
+    def request_exchange_token(name:, audience:, additional_scopes:, subject_token:, destination: nil)
       params = {
```
Comment (pepicrft) [inline lib/shopify_cli/identity_auth.rb:226]: @leleabhinav are these the right parameters to pass when exchanging the partners' token to generate one that we can use from the CLI?

**Comment by pepicrft** [comment]
This is no longer needed because [we've changed](https://github.com/Shopify/identity/pull/15984) the logic in Identity to already return a token that has the CLI as the audience.

File: test/shopify-cli/identity_auth/env_auth_token_test.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
+require "test_helper"
+
+module ShopifyCLI
+  class IdentityAuth
+    class EnvAuthTokenTest < MiniTest::Test
+      def test_partners_token_present
+        # Given
+        Environment.expects(:auth_token).returns("token")
```
Comment (gonzaloriestra) [inline test/shopify-cli/identity_auth/env_auth_token_test.rb:8]: The Given part should not have expectations, but stubs.

```suggestion
        Environment.stubs(:auth_token).returns("token")
```

File: test/shopify-cli/identity_auth/env_auth_token_test.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        # When
+        got = EnvAuthToken.partners_token_present?
+
+        # Then
+        assert got
+      end
+
+      def test_exchanged_partners_token
+        # Given
+        EnvAuthToken.expects(:exchanged_partners_token=).with do |token|
```
Comment (gonzaloriestra) [inline test/shopify-cli/identity_auth/env_auth_token_test.rb:19]:
```suggestion
        EnvAuthToken.stubs(:exchanged_partners_token=).with do |token|
```

File: test/shopify-cli/identity_auth/env_auth_token_test.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        end
+
+        # Then
+        assert_equal "existing", got
+      end
+
+      def test_exchanged_partners_token_exchanges_the_token_if_the_existing_has_expired
+        # Given
+        expires_at = Time.now.to_i - 3000
+        EnvAuthToken.expects(:exchanged_partners_token=).with do |token|
```
Comment (gonzaloriestra) [inline test/shopify-cli/identity_auth/env_auth_token_test.rb:54]:
```suggestion
        EnvAuthToken.stubs(:exchanged_partners_token=).with do |token|
```

File: test/shopify-cli/identity_auth_test.rb [hunk 5]
```diff
@@ -176,6 +176,23 @@ def test_timeout_waiting_for_resp
...
 
+    def test_fetch_or_auth_partners_token_when_environment_auth_token
+      # Given
+      env_auth_token = "token"
+      exchanged_token = "exchanged_token"
+      IdentityAuth::EnvAuthToken.stubs(:partners_token_present?).returns(true)
+      IdentityAuth::EnvAuthToken.stubs(:fetch_exchanged_partners_token)
+        .yields(env_auth_token)
+        .returns(exchanged_token)
+      IdentityAuth.any_instance.expects(:exchange_partners_auth_token).returns(exchanged_token)
```
Comment (gonzaloriestra) [inline test/shopify-cli/identity_auth_test.rb:187]:
```suggestion
      IdentityAuth.any_instance.stubs(:exchange_partners_auth_token).returns(exchanged_token)
```

**Review by gonzaloriestra — APPROVED** [review]
Great! 👏 

Just some nitpicking suggestions.

**Pull request closed without merging**